tephra-core/src/main/java/co/cask/tephra/runtime/ZKModule.java

```diff
@@ -52,16 +62,68 @@ private ZKClientService provideZKClientService(Configuration conf) {
       zkStr = conf.get(TxConstants.HBase.ZOOKEEPER_QUORUM);
     }
 
-    return ZKClientServices.delegate(
-      ZKClients.reWatchOnExpire(
-        ZKClients.retryOnFailure(
-          ZKClientService.Builder.of(zkStr)
-            .setSessionTimeout(conf.getInt(TxConstants.HBase.ZK_SESSION_TIMEOUT,
-                TxConstants.HBase.DEFAULT_ZK_SESSION_TIMEOUT))
-            .build(),
-          RetryStrategies.exponentialDelay(500, 2000, TimeUnit.MILLISECONDS)
-        )
-      )
-    );
+    ZKClientService zkClientService = ZKClientService.Builder.of(zkStr)
+      .setSessionTimeout(conf.getInt(TxConstants.HBase.ZK_SESSION_TIMEOUT,
+                                     TxConstants.HBase.DEFAULT_ZK_SESSION_TIMEOUT)).build();
+
+    ZKClient zkClient = ZKClients.retryOnFailure(zkClientService,
+                                                 RetryStrategies.exponentialDelay(500, 2000, TimeUnit.MILLISECONDS));
+
+    final ZKClient client = ZKClients.reWatchOnExpire(zkClient);
+    return new ForwardingZKClientService(zkClientService) {
```

Comment:
Copied `ForwardingZKClientService` to Tephra which now extends from Tephra's `ZKCLientService`.